Fix missing parameter check in at::batch_norm

hipudding opened the pull request:
Input tensor valid checking are defined in
torch.nn.functional::batch_norm, if at::batch_norm is called, it may access illegal memory addresses.

Move valid checking to inner function to avoid memory corruption.

Fixes #117245


cc @albanD @mruberry @jbschlosser @walterddr @mikaylagawarecki

hipudding committed:
Fix missing parameter check in at::Batch_norm

Input tensor valid checking are defined in
torch.nn.functional::batch_norm, if at::batch_norm is called, it
may access illegal memory addresses.

Move valid checking to inner function to avoid memory corruption.

issue #117245

albanD reviewed (commented):
I don't remember where else the batch_norm() call can end up. We definitely want to make sure we cover all cases. @andrewor14 do you know off the top of your head if this move is enough or we need these checks somewhere else?

Also I guess these checks can also be removed from the python code that calls torch.batch_norm?

andrewor14 commented on aten/src/ATen/native/Normalization.cpp:
```diff
@@ -485,6 +485,23 @@ std::tuple<Tensor, Tensor, Tensor> batch_norm_backward_cpu_template(
 std::tuple<Tensor, Tensor, Tensor, Tensor, int64_t> _batch_norm_impl_index(
```
Moving forward, this won't be called after batch norm consolidation. I would recommend moving this check to the `batch_norm` function itself

andrewor14 commented:
> I don't remember where else the batch_norm() call can end up. We definitely want to make sure we cover all cases. @andrewor14 do you know off the top of your head if this move is enough or we need these checks somewhere else?
> 
> Also I guess these checks can also be removed from the python code that calls torch.batch_norm?

These checks should be enough for the current state, since both eager and compile/export go through this op. This won't be true after batch norm consolidation, however. A better place to do these checks would be in `aten.batch_norm` itself; that will always be the common aten op for both flows

andrewor14 commented:
> I don't remember where else the batch_norm() call can end up. We definitely want to make sure we cover all cases. @andrewor14 do you know off the top of your head if this move is enough or we need these checks somewhere else?
> 
> Also I guess these checks can also be removed from the python code that calls torch.batch_norm?

These checks should be enough for the current state, since both eager and compile/export go through this op. However, this will no longer be true after batch norm consolidation. A better place to do these checks would be in `aten.batch_norm` itself; that will always be the common aten op for both flows

hipudding commented:
Thanks @albanD @andrewor14 for your review. I was exciting that this is my first contribute to pytorch.

I followed call stack of batch_norm, there's a lot dispatcher stack, and I thought _batch_norm_impl_index is the first not generate function or dispatch function. I found aten.batch_norm function are generated by gen.sh, do you mean I should to change something in code generation?

albanD commented:
From the CI signal, it does seem like these tests are too restrictive though :/

andrewor14 commented:
> Thanks @albanD @andrewor14 for your review. I was exciting that this is my first contribute to pytorch.
> 
> I followed call stack of batch_norm, there's a lot dispatcher stack, and I thought _batch_norm_impl_index is the first not generate function or dispatch function. I found aten.batch_norm function are generated by gen.sh, do you mean I should to change something in code generation or something in code template?

I don't think `aten.batch_norm` is generated? Can you add the check here? https://github.com/pytorch/pytorch/blob/main/aten/src/ATen/native/Normalization.cpp#L640

hipudding commented:
> From the CI signal, it does seem like these tests are too restrictive though :/ 
> 
> 

Thanks. I will check it.

hipudding commented:
> > Thanks @albanD @andrewor14 for your review. I was exciting that this is my first contribute to pytorch.
> 
> > 
> 
> > I followed call stack of batch_norm, there's a lot dispatcher stack, and I thought _batch_norm_impl_index is the first not generate function or dispatch function. I found aten.batch_norm function are generated by gen.sh, do you mean I should to change something in code generation or something in code template?
> 
> 
> 
> I don't think `aten.batch_norm` is generated? Can you add the check here? https://github.com/pytorch/pytorch/blob/main/aten/src/ATen/native/Normalization.cpp#L640

Got it. Thanks.

hipudding committed:
Move batch norm check to batch_norm function

hipudding committed:
Remove 0-dim input tensor in batch_norm test

batch_norm need at least 2 input dimensions, 0 dim input test case
is not suitable for this function.

albanD commented on torch/testing/_internal/common_methods_invocations.py:
```diff
@@ -413,8 +413,6 @@ def sample_inputs_batch_norm(op_info, device, dtype, requires_grad, **kwargs):
         ((3, 2, 4), {'training': False, 'momentum': -1.2}),
         ((3, 1), {'training': True, 'momentum': 0.0}),
-        ((0,), {'training': True}),
-        ((0,), {'training': False}),
```
Why are you removing these test cases?

hipudding replied:
According to the content of this PR modification, batch_norm cannot accept tensors less than 2dims, so these two input data will cause many test cases failed.
I have already add a testcase to cover tensor less than 2dims.

albanD replied:
The thing is that these two tests used to pass. So if we're saying they're not passing anymore, this would be BC-breaking.
I don't think we should change the behavior of batchnorm in any way here.

We would need to make sure the test only prevent the case that segfault but still allow anything that used to be valid before.

hipudding replied:
Thank you for your reply. But the fact is the only thing I have done in this PR is move input parameters check from outer function to barch_norm itself, because some times, this parameter check is not called(the check is not in every call stack).  I the past, these two testcase can pass because batch_norm can handle empty tensor(it will return another empty tensor as the result), but even so, some code has already visit invalid memory(such as here: https://github.com/pytorch/pytorch/blob/677e67c3998df11bee0d371f9f6f9482c9e4c29e/aten/src/ATen/native/Normalization.cpp#L495, if input only as 1 dim, input.sym_sizes()[1] is not a valid address). That why I create this PR to prevent input less than 2 dims.

For these two test cases, input only has 1 dim, it will checked by new added parameter check, So I remove them. 

Do you have any good suggestions to keep these two test cases but fix this issue? Thanks.

albanD replied:
Someone would need to dive into details of what is expected or not for the python side of things but the sample just above this one is `((3, 1), {'training': True, 'momentum': 0.0}),` which would mean a 2D input IIRC.

I'm not saying this is bug free but the key expectation for this code is stability. So I do not want to make any change that might change what works or not without understanding in details the state of what is expected.
Sorry for dragging this PR but this is a very high profile function and we definitely don't want to break anything that is working today.

hipudding replied:
Thanks. I got it. I will try to not break anything and fix this issue.
